File: android/app/src/main/kotlin/pt/ualg/miaugenda/data/remote/ShiftApi.kt
package pt.ualg.miaugenda.data.remote

import pt.ualg.miaugenda.data.model.Shift
import pt.ualg.miaugenda.data.model.ShiftType
import pt.ualg.miaugenda.data.model.CreateShiftRequest
import pt.ualg.miaugenda.data.model.UpdateShiftRequest
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

interface ShiftApi {

    @GET("/shifts")
    suspend fun getShifts(
        @Query("week") week: String? = null,
        @Query("month") month: String? = null
    ): Response<List<Shift>>

    @GET("/shifts/me")
    suspend fun getMyShifts(
        @Query("week") week: String? = null,
        @Query("month") month: String? = null
    ): Response<List<Shift>>

    @POST("/shifts")
    suspend fun createShift(
        @Body request: CreateShiftRequest
    ): Response<Shift>

    @PATCH("/shifts/{id}")
    suspend fun updateShift(
        @Path("id") id: Int,
        @Body request: UpdateShiftRequest
    ): Response<Shift>

    @DELETE("/shifts/{id}")
    suspend fun deleteShift(
        @Path("id") id: Int
    ): Response<Unit>

    @GET("/shift-types")
    suspend fun getShiftTypes(): Response<List<ShiftType>>
}

File: desktop/src/api/shiftApi.ts
import { apiClient } from "./apiClient";
import type {
  CreateShiftRequest,
  Shift,
  ShiftType,
  UpdateShiftRequest
} from "../models/shift";

export interface ShiftFilters {
  week?: string | null;
  month?: string | null;
  userId?: number | null;
}

export async function getShifts(filters: ShiftFilters = {}): Promise<Shift[]> {
  const response = await apiClient.get<Shift[]>("/shifts", {
    params: {
      week: filters.week ?? undefined,
      month: filters.month ?? undefined,
      userId: filters.userId ?? undefined
    }
  });

  return response.data;
}

export async function getMyShifts(
  filters: Omit<ShiftFilters, "userId"> = {}
): Promise<Shift[]> {
  const response = await apiClient.get<Shift[]>("/shifts/me", {
    params: {
      week: filters.week ?? undefined,
      month: filters.month ?? undefined
    }
  });

  return response.data;
}

export async function createShift(request: CreateShiftRequest): Promise<Shift> {
  const response = await apiClient.post<Shift>("/shifts", request);
  return response.data;
}

export async function updateShift(
  id: number,
  request: UpdateShiftRequest
): Promise<Shift> {
  const response = await apiClient.patch<Shift>(`/shifts/${id}`, request);
  return response.data;
}

export async function deleteShift(id: number): Promise<void> {
  await apiClient.delete(`/shifts/${id}`);
}

export async function getShiftTypes(): Promise<ShiftType[]> {
  const response = await apiClient.get<ShiftType[]>("/shift-types");
  return response.data;
}